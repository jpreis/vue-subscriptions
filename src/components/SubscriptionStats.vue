<template>
  <p class="subscription-stats">
    <template v-if="subscriptionCount > 0">
      I currently have
      <strong>{{ activeSubscriptionCount }} active {{ subscriptionString }}</strong>. <br> That's
      <strong>{{ totalMonthlyCosts | dollars }} a month </strong> and
      <strong>{{ totalYearlyCosts | dollars }} a year</strong>. </template>
    <template v-else>
      No subscriptions yet...
    </template>
  </p>

</template>

<script>
import { dollars } from "@/filter/CurrencyFilter";

export default {
  props: {
    subscriptions: {
      required: true
    }
  },
  computed: {
    subscriptionCount() {
      return this.subscriptions.length;
    },
    activeSubscriptionCount() {
      return this.subscriptions.filter(subscription => subscription.active)
        .length;
    },
    totalMonthlyCosts() {
      return this.subscriptions
        .filter(subscription => subscription.active)
        .reduce((sum, subscription) => {
          return sum + subscription.costs;
        }, 0);
    },
    totalYearlyCosts() {
      return this.totalMonthlyCosts * 12;
    },
    subscriptionString() {
      return `subscription${this.activeSubscriptionCount !== 1 ? "s" : ""}`;
    }
  },
  filters: {
    dollars
  }
};
</script>

<style scoped>
.subscription-stats {
  margin-top: 2rem;
}
</style>