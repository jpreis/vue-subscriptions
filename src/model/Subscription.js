export default class Subscription {
  static nextId = 0;

  constructor(name, costs, active = true) {
    this.id = Subscription.nextId++;
    this.name = name;
    this.costs = costs;
    this.active = active;
  }
}
