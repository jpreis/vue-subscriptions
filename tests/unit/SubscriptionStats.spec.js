import { shallowMount } from "@vue/test-utils";
import SubscriptionStats from "@/components/SubscriptionStats.vue";
import Subscription from "@/model/Subscription";
import Vue from "vue";

describe("SubscriptionStats.vue", () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(SubscriptionStats, {
      propsData: {
        subscriptions: []
      }
    });
  });

  it("has the expected HTML structure", () => {
    expect(wrapper.element).toMatchSnapshot();

    wrapper.setProps({
      subscriptions: [
        new Subscription("Sub 1", 10),
        new Subscription("Sub 2", 20)
      ]
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
