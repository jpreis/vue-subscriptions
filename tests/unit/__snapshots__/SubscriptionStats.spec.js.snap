// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubscriptionStats.vue has the expected HTML structure 1`] = `
<p
  class="subscription-stats"
>
  
    No subscriptions yet...
  
</p>
`;

exports[`SubscriptionStats.vue has the expected HTML structure 2`] = `
<p
  class="subscription-stats"
>
  
    I currently have
    
  <strong>
    2 active subscriptions
  </strong>
  . 
  <br />
   That's
    
  <strong>
    $30.00 a month 
  </strong>
   and
    
  <strong>
    $360.00 a year
  </strong>
  . 
</p>
`;
